### PR TITLE
 Prototyping ksr restart resync - exploring building blocks DO NOT MERGE

### PR DESCRIPTION
**Pull request by jmedved**
Exploring how to implement ksr resync when ksr restarts; we will need to do something like:
- check that resync with k8s is finished
- read etcd state
- List k8s cache & mark sweep what's in etcs

We must avoid churning staeady state, i.e. if a value in etcd is not changing, we do not write. We need to update changed values, add new values and delete values no longer present in etcd. And we need to hold off any add/delete/update events until we're done.

### CODE REVIEW

**Comment by jmedved** [comment]
Rasto, Lukas, Milan - can you guys please review, and if it's ok, merge? I only refactored the service reflector so far, the other ones stiff needs to be done. 

I have another question - how do we detect Etcd failure, and how do we recover from that?

File: plugins/ksr/ksr_reflector.go [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+
+	kr.dsSynced = true
+}
+
+func (kr *Reflector) syncDataStore(dsItems DsItems, oc K8sToProtoConverter) {
+	for {
+		if kr.k8sController.HasSynced() {
+			break
+		}
+		time.Sleep(1 * time.Second)
```
Comment (rastislavszabo) [inline plugins/ksr/ksr_reflector.go]: I would put some smaller sleep time here (100ms ?)

**Comment by lukasmacko** [comment]
ad etcd failure detection, 
currently it is possible to detect etcd connection failure using health/statuscheck plugin. It allows to retrieve current state of registered plugins and check if the etcd connectivity is ok. In order to recover properly I would suggest to add watcher API to statuscheck, that would notify subscribers on state change. Re-established etcd connection after failure would trigger resync.

**Comment by jmedved** [comment]
Lukas, do you have an example of how to use the health/statuscheck plugin? Please point me to the code. 

In the meantime, can you guys please merge this patch, i would like to continue the work on a new patch:
- Add tests
- Migrate all other Reflectors onto the new structure (I started with Endpoints already, the Reflecto code is cleaned up quite a bit)

**Comment by lukasmacko** [comment]
with the following patch applied to cn-infra
[statuscheckpatch.txt](https://github.com/contiv/vpp/files/1614524/statuscheck.txt)
you can read status of etcd using datasync example

**Pull request merged**